test/x3/only_1_xform.cpp

```diff
@@ -0,0 +1,66 @@
+/*=============================================================================
+    Copyright (c) 2019 Larry Evans
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+=============================================================================*/
+//Purpose:
+//  test that only 1 attribute transformation occurs.
+//Method:
+//  specialize transform_attribute<int,int> to
+//  ++attribute.  If the transform done more than once
+//  this should be evident in the resulting value.
+//===============
+#include <boost/detail/lightweight_test.hpp>
+#include <boost/spirit/home/x3.hpp>
+namespace x3=boost::spirit::x3;
+
+x3::rule<class trace_xform_id,int> xform_rule="xform_rule";
+namespace boost { namespace spirit { namespace x3
+{
+    template <>
+    struct transform_attribute<int,int>
+    {
+        typedef int type;
+
+        static type pre(int&attr) 
+        { 
+            return attr;
+        }
+
+        static void post(int& val, type&& attr)
+        {
+            val=attr+1;//non-default transform
+        }
+    };
+}}}
+auto xform_rule_def=x3::int_;
```

Comment:
Test nested/recursive rules too.

Reply:
On 3/3/19 9:43 AM, Nikita Kniazev [wrote](https://github.com/boostorg/spirit/pull/479#issuecomment-469034889):

> You wrote it will simplify the code, but it is not, even
> opposite. 

So, let me see, you think [control coupling] is good, or
that it's just not as bad as the solution proposed in [PR479]?

> It makes debugging more painful, control flow jumps three
> times more than currently. 

Again, where are these unneeded control flow jumps.  How did
you calculate this "three times more than currently"?  I
keep asking for specifics and you keep giving me obscure
responses.

OK, I suppose I'll have to, once again, guess your meaning :(

Do you mean the use of the `parser_f` lambdas
[here](https://github.com/boostorg/spirit/pull/479/files#diff-4af6d86666f07de96ec4a2cbd58d767aR68)
and
[here](https://github.com/boostorg/spirit/pull/479/files#diff-4af6d86666f07de96ec4a2cbd58d767aR145)
is what's obfuscated?  And the unneeded control flow jumps
are jumps into and out of these `lambdas`?  Well the obvious
solution is to simply manually inline the code from the body
of the `lambda`, but that's the job for the compiler, not
the code maintainer.  Hence, I doubt that's the code your
talking about, but who knows.

OK, now it's your turn.  If the above is not what you mean
by obfuscated code, then where is the obfuscated code, and,
again, be specific so people don't have to keep guessing
what you mean.

> I can link Wikipedia with the meaning of obfuscation,
> shortly "obfuscation is the deliberate act of creating
> source or machine code that is difficult for humans to
> understand".

Of course I understand what obfuscation means.  What I don't
understand is why you think this "wherever" code is
obfuscated.  It's simple, show me the allegedly obfuscated
code and maybe I can understand what you mean or how to
clarify it.

>> Aren't there are already tests for nested/recursive
>> rules?  This PR's code has passed all tests according to
>> what I see in [PR479].  Hence, if there are no such tests
>> specific to nested/recursive rules, then that should be
>> another PR.
 
> You added a test to ensure there is no transformations on
> the same type is called, I previously pointed you that it
> is happening in nested/recursive rules and rule
> definitions already, but you have chosen to ignore it. 

I did *not* ignore it, I didn't understand it. I also
explicity said that
[here](https://github.com/boostorg/spirit/issues/459#issuecomment-468960665):
with the specific paragraph:

>> Sorry, I see no connection between nested rules and
>> redundant transform calls for a single occurrence of a
>> rule. Please clarify.

So if anyone is ignoring something, its you.

> Fix this and also remove the pass-through transformation,

The pass-through transformation?  What specifically are you
talking about.  By specific, I mean file and line number, as
I requested before but which you ignored.

> this will be at least one simplifying change.

Maybe, but I can't tell since you're, again, being vague.

[control coupling]: https://en.m.wikipedia.org/wiki/Coupling_(computer_programming)
[PR479]: https://github.com/boostorg/spirit/pull/479

Reply:
Please, stop replying under unrelated comments, open GitHub, look what you are doing, this is not a fist time.